### PR TITLE
Tooltip: Improved Heatmap tooltip

### LINKED ISSUE #75843: Tooltip: Improved Heatmap tooltip
[Figma link](https://www.figma.com/file/tW8PMdzLb7hYYyBpb2xaYR/Universal-Tooltip?type=design&node-id=238-2638&mode=design&t=sDsktJ1imHOLLtdO-0)

Screenshot 2023-10-02 at 1 38 39 PM

Screenshot 2023-10-02 at 1 38 54 PM

### PR DESCRIPTION
**Pull request by adela-almasan**
Heatmap tooltip using `TooltipPlugin2`

Blocked by
[Dashboard: Overflow and z-index adjustments](https://github.com/grafana/grafana/pull/75597) ✅ 
[Tooltip: VizTooltip components #75794](https://github.com/grafana/grafana/pull/75794) ✅ 

https://github.com/grafana/grafana/assets/88068998/fa68285e-22b0-4e61-8d68-5ca8b0e161ca

Fixes #75843 
Fixes #76131

Test file
[native_histogram_heatmap.json.txt](https://github.com/grafana/grafana/files/12874312/debug-Native-2023-10-06.21.52.20.json.txt)

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: public/app/plugins/panel/heatmap/tooltip/HeaderLabel.tsx
Comment (adela-almasan) [inline public/app/plugins/panel/heatmap/tooltip/HeaderLabel.tsx:1]: most of what's under heatmap/tooltip are part of a https://github.com/grafana/grafana/pull/75794. Will rebase after it gets merged.

File: public/app/plugins/panel/heatmap/HeatmapPanel.tsx [hunk 1]
```diff
@@ -160,22 +129,11 @@ export const HeatmapPanel = ({
...
-    if (hover && info.heatmap.fields && hover.seriesIdx === 1) {
-      hoverValue = countField.values[hover.dataIdx];
-    }
-
     return (
       <VizLayout.Legend placement="bottom" maxHeight="20%">
         <div className={styles.colorScaleWrapper}>
           <ColorScale
-            hoverValue={hoverValue}
+            // hoverValue={hoverValue}
```
Comment (nmarrs) [inline public/app/plugins/panel/heatmap/HeatmapPanel.tsx:136]: do we have a follow-up GH issue to track implementing hover value functionality?

File: public/app/plugins/panel/heatmap/module.tsx [hunk 2]
```diff
@@ -415,3 +423,4 @@ export const plugin = new PanelPlugin<Options, GraphFieldConfig>(HeatmapPanel)
     });
   })
   .setSuggestionsSupplier(new HeatmapSuggestionsSupplier());
+// .setDataSupport({ annotations: false });
```
Comment (nmarrs) [inline public/app/plugins/panel/heatmap/module.tsx:426]: Is this still necessary?

```suggestion
```

File: public/app/plugins/panel/heatmap/tooltip/HeatmapTooltip.tsx [hunk 3]
```diff
@@ -231,59 +262,110 @@ const HeatmapHoverCell = ({ data, hover, showHistogram, scopedVars, replaceVars
...
-      )}
-    </>
+      );
+    }
+
+    return null;
+  };
+
+  // @TODO remove this when adding annotations support
+  canAnnotate = false;
```
Comment (nmarrs) [inline public/app/plugins/panel/heatmap/tooltip/HeatmapTooltip.tsx:349]: will initial version of heatmap tooltip support annotations?

File: public/app/plugins/panel/heatmap/tooltip/tooltipUtils.ts [hunk 4]
```diff
@@ -0,0 +1,94 @@
+import { Field, FieldType, formattedValueToString } from '@grafana/data';
```
Comment (nmarrs) [inline public/app/plugins/panel/heatmap/tooltip/tooltipUtils.ts:1]: nit: but we probably could rename this file to just `utils` given it is in the `tooltip/` directory

**Review by nmarrs — COMMENTED** [review]
Will take another look when https://github.com/grafana/grafana/pull/75794 is merged - overall this is looking great!

File: packages/grafana-ui/src/components/uPlot/plugins/index.ts [hunk 5]
```diff
@@ -1,4 +1,5 @@
 export { ZoomPlugin } from './ZoomPlugin';
 export { TooltipPlugin } from './TooltipPlugin';
 export { TooltipPlugin2 } from './TooltipPlugin2';
+export { ZoomXPlugin } from './ZoomXPlugin';
```
[the remark this answers is not in the document]
Comment (adela-almasan) [reply packages/grafana-ui/src/components/uPlot/plugins/index.ts:4]: yes.. we'll do a cleanup at the very end..

File: public/app/plugins/panel/heatmap/HeatmapPanel.tsx (hunk 1 above)
Comment (adela-almasan) [reply public/app/plugins/panel/heatmap/HeatmapPanel.tsx:136]: Here: https://github.com/grafana/grafana/issues/75879

File: public/app/plugins/panel/heatmap/tooltip/HeatmapTooltip.tsx (hunk 3 above)
Comment (nmarrs) [reply public/app/plugins/panel/heatmap/tooltip/HeatmapTooltip.tsx:350]: jk - heatmaps dont currently support annotation if I remember correctly so disregard this question 😬

**Review by imatwawana — APPROVED** [review]
LGTM!

File: .betterer.results [hunk 6]
```diff
@@ -7427,6 +7418,13 @@ exports[`better eslint`] = {
       [0, 0, 0, "Do not use any type assertions.", "0"],
       [0, 0, 0, "Unexpected any. Specify a different type.", "1"]
     ],
+    "public/app/plugins/panel/heatmap/tooltip/utils.ts:5381": [
```
Comment (adela-almasan) [inline .betterer.results:7421]: I will try to clean up the anys..

File: public/app/plugins/panel/heatmap/tooltip/utils.ts [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+      break;
+    case 'count':
+      field = isSparse ? data?.fields.find((field) => field.name === 'count') : data?.fields[2];
+      break;
+  }
+
+  return field;
+};
+
+export const inferSparseDataBucketSizes = (data: HeatmapData, xVals: any[], yVals: any[]): BucketSizes => {
```
Comment (adela-almasan) [inline public/app/plugins/panel/heatmap/tooltip/utils.ts:178]: I'm not completely sure these are correct. cc @leeoniya

**Comment by nmarrs** [comment]
We are considering adding a FF for tooltip stuff to go behind- with that in place and this behind that new FF this should be good for final review / merge :)

File: public/app/plugins/panel/heatmap/module.tsx [hunk 8]
```diff
@@ -406,6 +406,14 @@ export const plugin = new PanelPlugin<Options, GraphFieldConfig>(HeatmapPanel)
       showIf: (opts) => opts.tooltip.show,
     });
 
+    builder.addBooleanSwitch({
```
Comment (nmarrs) [inline public/app/plugins/panel/heatmap/module.tsx:409]: We should hide this behind FF :)

File: public/app/plugins/panel/heatmap/HeatmapPanel.tsx (hunk 1 above)
Comment (nmarrs) [reply public/app/plugins/panel/heatmap/HeatmapPanel.tsx:180]: Probably should remove legend changes so that we don't regress current legend behavior

**Review by nmarrs — COMMENTED** [review]
mob code review 😬

File: public/app/plugins/panel/heatmap/HeatmapPanel.tsx [hunk 9]
```diff
@@ -218,36 +226,66 @@ export const HeatmapPanel = ({
         {(vizWidth: number, vizHeight: number) => (
           <UPlotChart config={builder} data={facets as any} width={vizWidth} height={vizHeight}>
             {/*children ? children(config, alignedFrame) : null*/}
+            <ZoomPlugin
```
Comment (nmarrs) [inline public/app/plugins/panel/heatmap/HeatmapPanel.tsx:229]: are we confident releasing updated ZoomPlugin as part of this (i.e. not behind the FF?)

File: docs/sources/panels-visualizations/visualizations/heatmap/index.md [hunk 10]
```diff
@@ -114,6 +114,7 @@ Use these settings to refine your visualization.
 
 - **Show tooltip -** Show heatmap tooltip.
 - **Show Histogram -** Show a Y-axis histogram on the tooltip. A histogram represents the distribution of the bucket values for a specific timestamp.
+- **Show color scale -** Show a color scale on the tooltip. The color scale represents the mapping between bucket value and color.
```
Comment (nmarrs) [inline docs/sources/panels-visualizations/visualizations/heatmap/index.md:117]: This UI option will only show up if FF is enabled - maybe we should make note of that here (I think there are examples of this in the docs somewhere 🤔)

File: .betterer.results (hunk 6 above)
Comment (nmarrs) [reply .betterer.results:7041]: taking a 👀 at this :)

**Review by nmarrs — APPROVED** [review]
a few minor questions :) this is good overall though so I will approve and unblock you to merge when they are addressed 👍

File: public/app/features/visualization/data-hover/DataHoverView.tsx [hunk 11]
```diff
@@ -21,24 +21,28 @@ export interface Props {
...
   highlight: boolean;
 }
 
-export const DataHoverView = ({ data, rowIndex, columnIndex, sortOrder, mode, header = undefined }: Props) => {
-  const styles = useStyles2(getStyles);
-
-  if (!data || rowIndex == null) {
-    return null;
-  }
+export function getDisplayValuesAndLinks(
```
Comment (leeoniya) [inline public/app/features/visualization/data-hover/DataHoverView.tsx:35]: this just pulls out the logic from the the component body so we can reuse it.

File: public/app/features/visualization/data-hover/DataHoverView.tsx [hunk 12]
```diff
@@ -89,6 +93,24 @@ export const DataHoverView = ({ data, rowIndex, columnIndex, sortOrder, mode, he
     displayValues.sort((a, b) => arrayUtils.sortValues(sortOrder)(a.value, b.value));
   }
 
+  return { displayValues, links };
+}
+
+export const DataHoverView = ({ data, rowIndex, columnIndex, sortOrder, mode, header, padding = 0 }: Props) => {
```
Comment (leeoniya) [inline public/app/features/visualization/data-hover/DataHoverView.tsx:99]: only change here is optional `padding` param at end, rest of diff is just logic extraction to exportable fn.

File: public/app/plugins/panel/heatmap/HeatmapHoverViewOld.tsx
Comment (leeoniya) [inline public/app/plugins/panel/heatmap/HeatmapHoverViewOld.tsx:1]: this whole file is just a renamed of the unmodified `HeatmapHoverView.tsx`

**Pull request merged**